File: src/main/java/victor/training/cleancode/fp/LoanPattern.java
package victor.training.cleancode.fp;

import lombok.RequiredArgsConstructor;
import org.jooq.lambda.Unchecked;
import victor.training.cleancode.fp.support.OrderRepo;

import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.io.Writer;
import java.util.function.Consumer;

@RequiredArgsConstructor
class FileExportService {
   private final OrderRepo orderRepo;

   public void exportOrders(Consumer<Writer> action) throws IOException {
      File file = new File("target/orders.csv");
      System.out.println("Starting export into {} ... " + file.getAbsolutePath());
      long t0 = System.currentTimeMillis();
      try (Writer writer = new FileWriter(file)) {

         action.accept(writer); // transaction, connection, PDF, Excel

         System.out.println("Export DONE");
      } catch (Exception e) {
         System.out.println("Export FAILED: " + e); // TERROR-Driven Development
         // imagine... sendErrorEmail(e);
         throw e;
      } finally {
         long t1 = System.currentTimeMillis();
         System.out.println("Export completed in seconds: " + (t1 - t0) / 1000);
      }
   }
}


@RequiredArgsConstructor
public class LoanPattern {
  private final FileExportService fileExporterService;

  public void exportOrders() throws IOException {
    fileExporterService.exportOrders(writer ->{
//      writer.write("order_id;date\n");
//         orderRepo.findByActiveTrue()
//             .map(o -> o.id() + ";" + o.creationDate() + "\n")
//             .forEach(Unchecked.consumer(writer::write));
    });
  }

  public void exportUsers() {
    // TODO implement the export of users using *the same workflow* as for orders
  }
}

File: src/main/java/victor/training/cleancode/Loan_Pattern.kt
//package victor.training.fp
//
//import lombok.RequiredArgsConstructor
//import lombok.extern.slf4j.Slf4j
//import victor.training.cleancode.OrderRepo
//import java.io.File
//import java.io.FileWriter
//
//@Slf4j
//class FileExporter {
//    fun exportOrders(writeContent: (FileWriter) -> Unit, fileName: String) {
//        val file = File("target/$fileName")
//        println("Starting export into ${file.absolutePath} ...")
//        val t0 = System.currentTimeMillis()
//        try {
//            FileWriter(file).use { writer ->
//                writeContent(writer)
//                println("Export DONE")
//            }
//        } catch (e: Exception) {
//            println("Export FAILED!" + e) // TERROR-Driven Development
//            throw e
//        } finally {
//            val t1 = System.currentTimeMillis()
//            println("Export completed in " + (t1 - t0) / 1000 + " seconds ")
//        }
//    }
//}
//data class ContentWriters(private val orderRepo: OrderRepo) {
//
//     fun writeOrderContent(writer: FileWriter) {
//        writer.write("order_id;date\n")
//        orderRepo.findByActiveTrue()
//            .map { "${it.id};${it.creationDate}\n" }
//            .forEach { writer.write(it) }
//    }
//     fun writeUserContent(writer: FileWriter) {
//        writer.write("order_id;date\n")
////        orderRepo.findByActiveTrue()
////            .map { "${it.id};${it.creationDate}\n" }
////            .forEach { writer.write(it) }
//    }
//}
//
//@RequiredArgsConstructor
//internal class ExportService(private val fileExporter: FileExporter) {
//    fun exportOrders() {
//        fileExporter.exportOrders({ContentWriters().writeOrderContent(it)}, "orders.csv")
//    }
//
//    fun exportUsers() {
//        fileExporter.exportOrders({ContentWriters().writeUserContent(it)}, "users.csv")
//    }
//}